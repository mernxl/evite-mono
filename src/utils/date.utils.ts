import moment, { Moment } from 'moment';
import { pick } from 'ramda';

export interface ParsedUTCOffset {
  op: 'add' | 'sub';
  offset: string;
  hours: number;
  minutes: number;
  timezone: string;
}

const TimezoneFixedUTCMap: Record<string, string> = {
  'UTC-11': 'Pacific/Pago_Pago',
  'UTC-10': 'Pacific/Honolulu',
  'UTC-9': 'America/Anchorage',
  'UTC-8': 'America/Los_Angeles',
  'UTC-7': 'America/Denver',
  'UTC-6': 'America/Chicago',
  'UTC-5': 'America/Detroit',
  'UTC-4': 'America/Caracas',
  'UTC-3': 'America/Sao_Paulo',
  'UTC-2': 'Atlantic/South_Georgia',
  'UTC-1': 'Atlantic/Cape_Verde',
  'UTC+0': 'UTC',
  'UTC+1': 'Africa/Lagos',
  'UTC+2': 'Africa/Cairo',
  'UTC+3': 'Africa/Addis_Ababa',
  'UTC+4': 'Asia/Dubai',
  'UTC+5': 'Asia/Karachi',
  'UTC+6': 'Asia/Dhaka',
  'UTC+7': 'Asia/Bangkok',
  'UTC+8': 'Asia/Shanghai',
  'UTC+9': 'Asia/Seoul',
  'UTC+10': 'Asia/Brisbane',
  'UTC+11': 'Australia/Sydney',
  'UTC+12': 'Pacific/Norfolk',
  'UTC+13': 'Pacific/Tongatapu',
  'UTC+14': 'Pacific/Apia',
};

export const parseUTCOffset = (utcOffset = ''): ParsedUTCOffset => {
  const [hours, minutes] = utcOffset.split(':');

  let _hours;
  let op: ParsedUTCOffset['op'] = 'add';
  let offset = '+';

  if (hours[0] === '-' || hours[0] === '+') {
    if (hours[0] === '-') {
      op = 'sub';
      offset = '-';
    }
    _hours = parseInt(hours.substr(1)) || 0;
  } else {
    _hours = parseInt(hours) || 0;
  }

  const _minutes = (parseInt(minutes) || 0) % 60;

  _hours = _hours % 24;
  offset += _hours < 10 ? `0${_hours}` : _hours;
  offset += ':';
  offset += _minutes < 10 ? `0${_minutes}` : _minutes;

  return {
    op,
    offset,
    hours: _hours,
    minutes: _minutes,
    timezone: TimezoneFixedUTCMap[`UTC${op === 'add' ? '+' : '-'}${_hours}`] || 'UTC',
  };
};

export type UTCOffset = string | ParsedUTCOffset;

/**
 * This method assumes every input that every input is a local input
 * All it does is add to or subtract from
 *
 * NOTE: Only use when wanting to get the utc date from an offset of a local date
 *
 * We reverse ops in order to catch up with passed date, i.e.
 *
 * If we are requesting a +1 of 00:00, then its actually stored in system as 23:00
 * In the above case we are subtracting from the local time to match the appropriate utc
 * Hence the toUTC option for such cases
 *
 * In nutshell,
 *  toUTC=true takes from offset date to utc date
 *  toUTC=false takes from a utc date to offset date
 * The computation takes place with the utc date.
 */
export function applyUTCOffset(
  utcInput: Moment | Date | moment.MomentInput,
  utcOffset: UTCOffset = '+0',
  toUTC = false,
): Moment {
  if (typeof utcOffset === 'string') {
    utcOffset = parseUTCOffset(utcOffset);
  }

  let utcMoment: Moment;
  if (moment.isMoment(utcInput)) {
    utcMoment = utcInput as Moment;
  } else {
    utcMoment = moment(utcInput);
  }

  const duration = pick(['hours', 'minutes'], utcOffset);
  const op = toUTC ? (utcOffset.op === 'add' ? 'sub' : 'add') : utcOffset.op;

  if (op === 'add') {
    utcMoment = utcMoment.add(duration);
  } else {
    utcMoment = utcMoment.subtract(duration);
  }

  return utcMoment;
}
