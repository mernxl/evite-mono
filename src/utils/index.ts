export * from './date.utils';
export * from './docMeta.utils';
export * from './misc.utils';
export * from './mongoose';
export * from './pagination.utils';
export * from './password.utils';
export * from './validations';
