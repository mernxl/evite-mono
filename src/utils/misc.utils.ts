import { StatusCodes } from 'http-status-codes';

export interface IErrorResponse<T = string> {
  status: number;
  message: string;
  details: ValidationFieldErrors<T>;
}

/**
 * General Error Object, gets other props from Error which include
 *
 * name, message, stack
 */
export class ErrorResponse extends Error {
  constructor(
    public message = 'Internal Server Error',
    public status = StatusCodes.INTERNAL_SERVER_ERROR,
    public details?: ValidationFieldErrors<string> | unknown,
  ) {
    super(message);
    this.status = status || this.status;
  }
}

export type ValidationFieldErrors<T> = {
  [K in keyof T]?: string | ValidationFieldErrors<any>;
};

export class ValidationError<T> extends ErrorResponse {
  constructor(public fields: ValidationFieldErrors<T>, message = 'Input Validation Error') {
    super(message, StatusCodes.BAD_REQUEST, fields);
  }
}

export class AuthorizationError extends ErrorResponse {
  name = 'AuthorizationError';
  status = StatusCodes.UNAUTHORIZED;

  constructor(public message: string, public details?: unknown) {
    super(message, undefined, details);
  }
}

export class ForbiddenError extends ErrorResponse {
  name = 'ForbiddenError';
  status = StatusCodes.FORBIDDEN;

  constructor(public message: string, public details?: unknown) {
    super(message, undefined, details);
  }
}

/**
 * Since null is not allowed on OpenAPI, use this to represent null values on Spec
 */
export type NullValue = null;

export type InferMongooseProps<T> = {
  [K in keyof T]: T[K];
};

export function enumObjectValues(enumObject: Record<string, any>): string[] {
  const valueArray: string[] = [];
  if (!enumObject) {
    return valueArray;
  }

  for (const key in enumObject) {
    valueArray.push(enumObject[key]);
  }

  return valueArray;
}

export const capitalizeStr = (str: string, separator = ' '): string =>
  str
    .split(separator)
    .map((s) => s[0].toUpperCase() + s.slice(1).toLowerCase())
    .join(separator);
