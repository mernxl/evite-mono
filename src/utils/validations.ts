import moment from 'moment';

export const invariant = <T>(checkValue: T, errorMessage: string): void => {
  if (!checkValue) {
    throw Error(errorMessage);
  }
};

export type StaleConfig = {
  [key in moment.DurationInputArg2]?: number; // object moment will parse to date
};

export const getStaleTimestamp = (staleConfig: StaleConfig): number =>
  moment().add(staleConfig).unix();
export const getStaleDate = (staleConfig: StaleConfig): Date => moment().add(staleConfig).toDate();

export const isItemStale = (stale = 0): boolean => moment().unix() > stale;

export const isEmailValid = (email: string): boolean =>
  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(
    email,
  );

/**
 * Ensure all emails are lower cased
 */
export const sanitizeEmail = (email = ''): string => email.trim().toLowerCase();

/**
 * Use for sanitizing phones ensure we store numbers without +, -, (, ) .i.e \D (non-digits)
 */
export const sanitizePhone = (phone = ''): string => phone.trim().replace(/(?!#)\D/g, '');

/**
 * For escaping using input when using that input as regex.
 * It can cause errors when passed directly to example db.
 *
 * $& means the whole matched string
 */
export const escapeRegExp = (string = ''): string => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

/**
 * Remove diacritics from a word
 *
 * normalize()ing to NFD Unicode normal form decomposes combined graphemes into the combination of simple ones.
 * The è of Crème ends up expressed as e +  ̀.
 *
 * @see https://stackoverflow.com/a/37511463/7387312
 */
export const removeDiacritics = (string = ''): string =>
  string.normalize('NFD').replace(/[\u0300-\u036f]/g, '');

export const parseRawStrInputForMatching = (str: unknown): string =>
  removeDiacritics(
    // remove all white spaces
    String(str).replace(/\s/g, ''),
  ).toLowerCase();
