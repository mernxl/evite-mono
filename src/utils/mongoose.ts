import { ClientSession, Connection, Document, DocumentDefinition } from 'mongoose';

import { IDocMetaUserTimestamp, UpdatableDocMeta } from './docMeta.utils';
import { escapeRegExp } from './validations';

export type CompoundIndex<T extends Document> = { [key in keyof Omit<T, keyof Document>]?: number };

/**
 * Converts a sort string to an object
 * Format
 *  -field.nest field => { "field.nest" : -1, field: 1 }
 */
export const sortStringToObject = (sort: string): Record<string, number> => {
  const sortObj: Record<string, number> = {};
  for (const string of sort.split(/\s+/)) {
    if (string[0] === '-') {
      sortObj[string.substr(1)] = -1;
    } else {
      sortObj[string] = 1;
    }
  }
  return sortObj;
};

/**
 * Search a term by regex
 */
export const getSearchByTermRegex = (term: string): RegExp => new RegExp(escapeRegExp(term), 'i');
export const getMatchBySearchTermRegex = <DocType extends Record<string, any> = any>(
  fields: (keyof DocType)[],
  term: string,
): { $or: Record<keyof DocType, RegExp>[] } => {
  const _term = escapeRegExp(term);

  return {
    $or: fields.map(
      (field) => ({ [field]: new RegExp(_term, 'i') } as Record<keyof DocType, RegExp>),
    ),
  };
};

/**
 * Function used to wrap session aware methods, and returns the value returned within that function.
 * This function assumes all models belong to about same db, or you pass a db where it will create a
 * fresh session.
 *
 * @param config
 * @param method
 */
export const sessionAwareMethod = async <T>(
  config: { db: Connection; session?: ClientSession },
  method: (session: ClientSession) => Promise<T>,
): Promise<T> => {
  const session = config.session || (await config.db.startSession());

  let returnValue: T;

  if (session.inTransaction()) {
    returnValue = await method(session);
  } else {
    await session.withTransaction(async () => {
      returnValue = await method(session);
    });

    session.endSession();
  }

  return returnValue!;
};

/**
 * Get a lowercase snake_case collection name from a CamelCase modelName
 * @param modelName - Module name to create collection name from
 * @param postfix - postfix to affix on the the collection name before returning, like s
 */
export const modelNameToCollectionName = (modelName: string, postfix = ''): string => {
  let collectionName = '';

  for (let i = 0; i < modelName.length; i++) {
    if (modelName[i].match(/[A-Z]/)) {
      if (i > 0) {
        collectionName += `_${modelName[i].toLowerCase()}`;
      } else {
        // if first, no underscore
        collectionName += `${modelName[i].toLowerCase()}`;
      }
    } else {
      collectionName += modelName[i];
    }
  }

  return collectionName + postfix;
};

/**
 * For populating extra fields on mongoose documents
 *
 * @param doc - Document to populate from
 * @param fieldIdName - field to populate
 */
export const populateField = async <DocType extends Document, ReturnType extends Document>(
  doc: DocType,
  fieldIdName: string & keyof DocType,
): Promise<ReturnType | null> => {
  // to avoid `MongoError: Use of expired sessions is not permitted`,
  // we need to set the session to null
  doc.$session(null as any);

  const old = doc[fieldIdName];

  const populated = await doc.populate(fieldIdName);

  if (old === populated[fieldIdName]) {
    return null;
  }

  doc[fieldIdName] = old;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore baseView()
  return Promise.resolve(populated[fieldIdName]);
};

/**
 * Field must be defined, either it can be undefined not optional
 */
export type ModifiableFields<Type> = {
  [index in keyof Type]: Type[index] | undefined;
};

/**
 * Since tsoa does no support direct undefined as type, we are forces to pass
 * data received from tsoa to make its type as the new, so it is statically validated.
 */
export const isModifiableData = <Type>(data: Type): ModifiableFields<Required<Type>> => data as any;

/**
 * A typed function to show errors when updating documents with missed fields
 *
 * Picks up missing fields especially in mapping case
 * Finally performs an update of the document by calling set method
 *
 * NOTE: All undefined fields will be removed, so if you really need to pass undefined to a field,
 * do it manually or pass an alternative like '' for string, false for boolean, 0 for numbers
 */
export const updateDoc = <DocType extends Document, DocBaseType extends Record<string, any>>(
  doc: DocType,
  update: ModifiableFields<
    Omit<DocumentDefinition<DocBaseType>, keyof Omit<IDocMetaUserTimestamp, keyof UpdatableDocMeta>>
  >,
): void => {
  // clean up all undefined. If you need to set to undefined, do so manually
  const cloneUpdate: Record<string, any> = {};

  for (const updateKey in update) {
    if (typeof (update as any)[updateKey] !== 'undefined') {
      cloneUpdate[updateKey] = (update as any)[updateKey];
    }
  }

  doc.set(cloneUpdate);
};
