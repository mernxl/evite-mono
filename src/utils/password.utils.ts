import { includes } from 'ramda';

import { invariant } from './validations';

export const CharURegex = /[A-Z]/;
export const CharLRegex = /[a-z]/;
export const NumberRegex = /[0-9]/;
export const OthersRegex = /[!#$%&()*+\-/?@[\]^_{|}~]/;

// generate random numbers, excluding 10
export const generateOTPCode = (length = 5): string =>
  Array(length).fill(10).map(randomClamp).join('');

/**
 * Generate a random number, clamped at max
 * It does not include max, since its considering from 0 to (max-1)
 * @param max
 */
const randomClamp = (max: number) => Math.min(parseInt((Math.random() * 10).toFixed(0)), max - 1);

/**
 * Method to set a character at a position in a string
 */
const setCharAt = (str: string, index: number, chr: string) => {
  if (index > str.length - 1) return str;

  return str.substr(0, index) + chr + str.substr(index + 1);
};

/**
 * Password generator
 * Should generate passwords that contain at least one of the following ASCII character sets
 * A - Z
 * a - z
 * 0 - 9
 * !#$%&()*+-/?@[]^_{|}~ at least one, then powers of 8
 *
 * So minimum password length is 4
 */
export const generatePassword = (len = 8): string => {
  const s = '!#$%&()*+-/?@[]^_{|}~';

  invariant(len >= 4, 'Minimum password length is 4.');

  const getCharFromStart = (start: string) => {
    if (start === '!') {
      return s.charAt(randomClamp(s.length));
    }
    return String.fromCodePoint(Number(randomClamp(start === '0' ? 10 : 26) + start.charCodeAt(0)));
  };

  let sb = '';
  // take len * 2 to have multiple to pick from when shuffling
  for (let i = 0; i < len * 2; i += 3) {
    sb += getCharFromStart('0');
    sb += getCharFromStart('a');
    sb += getCharFromStart('A');
  }

  // at least 1 of other, then we may add  / 8
  for (let i = 0; i < Math.ceil(len / 8); i++) {
    sb += getCharFromStart('!');
  }

  // shuffle the letters
  for (let i = 0; i < sb.length; i++) {
    const r = randomClamp(sb.length);
    const temp = sb.charAt(i);

    sb = setCharAt(sb, i, sb.charAt(r));
    sb = setCharAt(sb, r, temp);
  }

  // ensure at least 1 of all
  const matches: [RegExpMatchArray | null, string][] = [
    [sb.substr(0, len).match(CharURegex), 'A'],
    [sb.substr(0, len).match(CharLRegex), 'a'],
    [sb.substr(0, len).match(NumberRegex), '0'],
    [sb.substr(0, len).match(OthersRegex), '!'],
  ];

  const matched = matches
    .map((match) => (match[0] ? match[0].index : null))
    .filter((index) => index !== null);

  // ensure we have one of each, avoid over writing already check positions
  for (let i = 0; i < matches.length; i++) {
    if (!matches[i][0]) {
      let availableSpace = i;

      for (let j = 0; j < matches.length; j++) {
        // if space not already used
        if (!includes(j, matched)) {
          availableSpace = j;
          matched.push(j);
          break;
        }
      }

      sb = setCharAt(sb, availableSpace, getCharFromStart(matches[i][1]));
    }
  }

  return sb.slice(0, len);
};
