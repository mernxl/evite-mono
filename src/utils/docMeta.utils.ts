/* eslint-disable @typescript-eslint/ban-types */
import { Document, Model, Query, Schema, Types } from 'mongoose';

export interface IDocMetaTimestamp {
  createdAt: Date;
  updatedAt: Date;
}

export interface IDocMetaUserAction {
  createdById: Types.ObjectId;
  updatedById?: Types.ObjectId;
}

export interface DocMetaUserAction {
  createdById: string;
  updatedById?: string;
}

export type IDocMetaUserTimestamp = IDocMetaTimestamp & IDocMetaUserAction;

export interface UpdatableDocMeta {
  updatedAt: Date | undefined;
  updatedById: Types.ObjectId;
}

export type DocMetaConfig = { userId: Types.ObjectId };

export const addDocMetaTimestampToSchema = <
  DocType extends Document & IDocMetaTimestamp,
  ModelType extends Model<DocType> = any,
>(
  schema: Schema<DocType, ModelType>,
): void => {
  schema.set('timestamps', true);
};

export const addDocMetaUserActionToSchema = <
  DocType extends IDocMetaUserAction,
  ModelType extends Model<DocType> = any,
>(
  schema: Schema<DocType, ModelType>,
): void => {
  schema.add({
    createdById: { type: Schema.Types.ObjectId, required: true, ref: 'User' },
    updatedById: { type: Schema.Types.ObjectId, ref: 'User' },
  } as any);
};

export const addDocMetaToSchema = <
  DocType extends Document & IDocMetaUserTimestamp,
  ModelType extends Model<DocType> = any,
>(
  schema: Schema<DocType, ModelType>,
): void => {
  addDocMetaTimestampToSchema(schema);
  addDocMetaUserActionToSchema(schema);
};

export const addDocMetaUserTimestamp = <T extends object>(
  object: T,
  userId: Types.ObjectId | string,
): T & IDocMetaUserTimestamp => {
  const date = new Date();
  const userOId = typeof userId === 'string' ? new Types.ObjectId(userId) : userId;

  return {
    ...object,
    createdAt: date,
    updatedAt: date,
    createdById: userOId,
    updatedById: userOId,
  };
};

export const addUpdateDocMetaUserTimestamp = <T extends object>(
  object: T,
  userId: Types.ObjectId | string,
): T & Required<Pick<IDocMetaUserTimestamp, keyof UpdatableDocMeta>> => ({
  ...object,
  updatedAt: new Date(),
  updatedById: typeof userId === 'string' ? new Types.ObjectId(userId) : userId,
});

export const updateDocMetaUserTimestamp = (
  doc: IDocMetaUserTimestamp,
  userId: Types.ObjectId | string,
): void => {
  const update = addUpdateDocMetaUserTimestamp({}, userId);

  doc.updatedAt = update.updatedAt;
  doc.updatedById = update.updatedById;
};

export const viewDocMetaTimeStamp = (doc: IDocMetaTimestamp): IDocMetaTimestamp => ({
  createdAt: doc.createdAt,
  updatedAt: doc.updatedAt,
});

export const viewDocMetaUserTimestamp = (
  doc: IDocMetaUserTimestamp,
): IDocMetaTimestamp & DocMetaUserAction => ({
  ...viewDocMetaTimeStamp(doc),
  createdById: doc.createdById.toHexString(),
  updatedById: doc.updatedById?.toHexString(),
});

export const scopeQueryToCreatedBy = <T, T2 extends Document>(
  query: Query<T, T2>,
  userId: Types.ObjectId | string,
): Query<T, T2> => query.where('createdById', userId);
