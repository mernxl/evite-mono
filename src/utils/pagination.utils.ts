import { Document, Model, Query, Schema } from 'mongoose';

/**
 * Page to get, pages start from 1 to infinity
 * @default 1
 * @minimum 1
 */
export type PageQueryParam = number;

/**`
 * Items Per page, maximum to get is 100, defaults is 30
 * @default 30
 * @maximum 100
 */
export type PerPageQueryParam = number;

export interface Page<T> {
  /**
   * Actual data as requested
   */
  data: T[];
  /**
   * Total Elements available in the query made.
   */
  total: number;
  /**
   * Items on each page
   */
  perPage: number;
  /**
   * Current page number
   */
  currentPage: number;

  /**
   * If it has next page from current page
   */
  hasNext: boolean;

  /**
   * If it has prev page from current page
   */
  hasPrev: boolean;

  /**
   * Link to next page
   */
  nextPage: string;
  /**
   * Link to previous page
   */
  prevPage: string;
  /**
   * Link to first page
   */
  firstPage: string;
  /**
   * Link to last page
   */
  lastPage: string;
}

export interface PageOptions {
  /**
   * Path will be concatenated on basePath passed when setting this plugin
   */
  path: string;
  page?: number;
  perPage?: number;
}

export interface WithPaginatePluginOptions {
  /**
   * basePath we are serving this app from, will be concatenated with a later passed path
   * on the paginate function
   */
  basePath: string;
}

export type DataMapFn<T extends Document, M = T> = (
  dataItem: T,
  index: number,
  array: T[],
) => M | Promise<M>;

export type ExecutableQuery<T> = { exec: () => Promise<T> | T };
export type PageQueryType<T> = [
  ExecutableQuery<number>,
  {
    (data: { skip: number; limit: number }): ExecutableQuery<T[]> | Promise<ExecutableQuery<T[]>>;
  },
];

export interface ModelWithPaginatePlugin {
  /**
   * A pagination handler that takes a query that returns an array and paginates it depending on the
   * pageOptions.
   *
   * This method is set on the schema using the WithPaginatePlugin Plugin
   *
   * @param query
   * @param pageOptions
   * @param dataMapFn A function to use in mapping the data or a string which identifies a method on
   * that `model` that can map the data. If undefined, we will just pass data as is.
   */
  paginate<T extends Document, M = T>(
    query: Query<T[], T> | PageQueryType<T>,
    pageOptions: PageOptions,
    dataMapFn?: DataMapFn<T, M> | string,
  ): Promise<Page<M>>;
}

export const WithPaginatePlugin = <
  DocType extends Document,
  ModelType extends Model<DocType> = any,
>(
  schema: Schema<DocType, ModelType>,
  options: WithPaginatePluginOptions,
): void => {
  const getLink = (path: string, page: number, perPage: number): string =>
    `${options.basePath}${path}?page=${page}&per_page=${perPage}`;

  schema.static(
    'paginate',
    async function (
      this: Model<any>,
      query: Query<any[], any> | PageQueryType<any>,
      pageOptions: PageOptions,
      dataMapFn?: DataMapFn<any> | string,
    ): Promise<Page<any>> {
      const { path } = pageOptions;
      const page = Math.max(1, pageOptions.page || 1);
      const perPage = Math.min(100, pageOptions.perPage || 30); // 0 becomes 30

      let promises: [ExecutableQuery<number>, ExecutableQuery<any[]>];

      if (Array.isArray(query)) {
        promises = [query[0], await query[1]({ skip: (page - 1) * perPage, limit: perPage })];
      } else {
        // lets clone this query and make our count query
        const countQuery = new (query.toConstructor())();

        // promises to count and data query
        promises = [countQuery.countDocuments(), query.skip((page - 1) * perPage).limit(perPage)];
      }

      const [count, data] = (await Promise.all(promises.map((q) => (q as any).exec()))) as [
        number,
        any[],
      ];

      let preparedData = [];
      if (dataMapFn) {
        let mapFn: DataMapFn<any> | undefined;

        if (typeof dataMapFn === 'string') {
          // get that method from the model
          if (typeof (this as any)[dataMapFn] === 'function') {
            mapFn = (this as any)[dataMapFn];
          }
        } else if (typeof dataMapFn === 'function') {
          mapFn = dataMapFn;
        }

        if (mapFn) {
          preparedData = await Promise.all(data.map(mapFn));
        }
      } else {
        preparedData = data;
      }

      const page_count = Math.max(1, Math.ceil(count / perPage));

      return {
        data: preparedData,
        total: count,
        perPage: perPage,
        currentPage: page,

        hasNext: page + 1 <= page_count,
        hasPrev: page - 1 <= page_count && page - 1 > 0,

        nextPage: getLink(path, page + 1, perPage),
        prevPage: getLink(path, page - 1, perPage),
        firstPage: getLink(path, 1, perPage),
        lastPage: getLink(path, page_count, perPage),
      };
    },
  );
};
