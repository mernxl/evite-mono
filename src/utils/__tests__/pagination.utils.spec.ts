import { Document, Model, Schema } from 'mongoose';

import dbConnection, {
  cleanMongooseConnections,
  waitForOpenConnection,
} from '../../config/mongoose';
import { ModelWithPaginatePlugin, WithPaginatePlugin } from '../pagination.utils';

interface IDummy extends Document {
  index: number;
}
interface IDummyModel extends Model<IDummy>, ModelWithPaginatePlugin {}

const DummySchema = new Schema<IDummy, IDummyModel>({ index: Number });
WithPaginatePlugin(DummySchema, { basePath: 'localhost' });

const DummyModel = dbConnection.model<IDummy, IDummyModel>('PaginationTest', DummySchema);

describe('pagination.utils', () => {
  beforeAll(async (done) => {
    await waitForOpenConnection(dbConnection);
    await dbConnection.dropDatabase();

    await DummyModel.create(
      Array(100)
        .fill(null)
        .map((value, index) => ({ index })),
    );

    done();
  });

  afterAll(async () => {
    await cleanMongooseConnections();
  });

  it('should fetch page data, with default perPage 30 and page 1', async () => {
    const page = await DummyModel.paginate(DummyModel.find(), { path: 'dummy' });

    expect(page.data).toBeTruthy();
    expect(page.currentPage).toEqual(1);
    expect(page.perPage).toEqual(30);
  });

  it('should fetch page with correct info', async () => {
    const { data, ...rest } = await DummyModel.paginate(DummyModel.find(), {
      path: '/dummy',
      perPage: 30,
      page: 1,
    });

    expect(data.length).toEqual(30);
    expect(rest).toMatchSnapshot();
  });

  it('should apply the mapping function', async () => {
    const { data } = await DummyModel.paginate(
      DummyModel.find().sort({ index: 1 }),
      {
        path: '/dummy',
        perPage: 30,
        page: 1,
      },
      (dataItem: IDummy) => dataItem.index,
    );

    expect(data[0]).toEqual(0);
    expect(data[20]).toEqual(20);
  });

  it('should apply filter correctly on count and items', async () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { data, ...rest } = await DummyModel.paginate(
      DummyModel.find({ index: { $lt: 25 } }),
      {
        path: '/dummy',
        perPage: 30,
        page: 1,
      },
      (dataItem: IDummy) => dataItem.index,
    );

    expect(rest).toMatchSnapshot();
  });
});
