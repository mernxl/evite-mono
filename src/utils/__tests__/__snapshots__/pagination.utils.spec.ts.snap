// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination.utils should apply filter correctly on count and items 1`] = `
Object {
  "currentPage": 1,
  "firstPage": "localhost/dummy?page=1&per_page=30",
  "hasNext": false,
  "hasPrev": false,
  "lastPage": "localhost/dummy?page=1&per_page=30",
  "nextPage": "localhost/dummy?page=2&per_page=30",
  "perPage": 30,
  "prevPage": "localhost/dummy?page=0&per_page=30",
  "total": 25,
}
`;

exports[`pagination.utils should fetch page with correct info 1`] = `
Object {
  "currentPage": 1,
  "firstPage": "localhost/dummy?page=1&per_page=30",
  "hasNext": true,
  "hasPrev": false,
  "lastPage": "localhost/dummy?page=4&per_page=30",
  "nextPage": "localhost/dummy?page=2&per_page=30",
  "perPage": 30,
  "prevPage": "localhost/dummy?page=0&per_page=30",
  "total": 100,
}
`;
